PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,336 @@
+        
+        self.originalPurchaseDate = subscriberData.originalPurchaseDate
+        self.firstSeen = subscriberData.firstSeen
+        self.originalAppUserId = subscriberData.originalAppUserId 
+        self.managementURL = subscriberData.managementURL
+        self.originalApplicationVersion = subscriberData.originalApplicationVersion
+
+        self.nonConsumablePurchases = Set(subscriberData.nonSubscriptions.keys)
+        self.nonSubscriptionTransactions = subscriberData.nonSubscriptionTransactions
+
+        self.entitlements = EntitlementInfos.init(entitlementsData: subscriberData.entitlements, purchasesData: subscriberData.allPurchases, dateFormatter: DateFormatter.dateFormatter, requestDate: requestDate)
+
+        self.expirationDatesByProduct = subscriberData.expirationDatesByProduct
+        self.purchaseDatesByProduct = subscriberData.purchaseDatesByProduct
+    }
+
+    // TODO after migration make this internal
+    @objc public class func currentSchemaVersion() -> String {
+        return "2"
+    }
+
+    /**
+     Get the expiration date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for product
+    
+     @return The expiration date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func expirationDate(forProductIdentifier: String) -> Date? {
+        // todo why is this ?? necessary
+        return expirationDatesByProduct[forProductIdentifier] ?? nil
+    }
+
+     /**
+     Get the latest purchase or renewal date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for subscription product
+    
+     @return The purchase date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func purchaseDate(forProductIdentifier: String) -> Date? {
+        return purchaseDatesByProduct[forProductIdentifier] ?? nil
+    }
+
+    /**
+     Get the expiration date for a given entitlement.
+    
+     @param forEntitlement The id of the entitlement.
+    
+     @return The expiration date for the passed in `entitlement`, can be `nil`
+     */
+    @objc public func expirationDate(forEntitlement: String) -> Date? {
+        return entitlements[forEntitlement]?.expirationDate
+    }
+
+    /**
+     Get the latest purchase or renewal date for a given entitlement identifier.
+    
+     @param forEntitlement Entitlement identifier for entitlement
+    
+     @return The purchase date for `entitlement`, `nil` if product never purchased
+     */
+    @objc public func purchaseDate(forEntitlement: String) -> Date? {
+        return entitlements[forEntitlement]?.latestPurchaseDate
+    }
+
+    // TODO why were we previously able to call with lowercase from swift?
+    // TODO after migration make this internal
+    @objc(JSONObject) public func jsonObject() -> [AnyHashable: Any] {
```

Comment:
This was uppercase `JSONObject` in objc before, but was called from swift with lowercase `jsonObject`...does anyone know how that ever worked? In either case, this should be internal, so we can migrate to lowercase post-migration...

Reply:
I think it's just the obj-c -> swift naming translation

Reply:
https://github.com/apple/swift-evolution/blob/master/proposals/0005-objective-c-name-translation.md

> Lowercase values: The Swift API Design Guidelines have non-type declarations lowercased. Lowercase non-prefixed values whenever they are imported, including enumerators (whether they end up in Swift as enum cases or option sets) and any other properties/functions (e.g., a property named URLHandler will be lowercased to urlHandler).

Reply:
are the keys not guaranteed to be strings? if they are, we should constrain it to `[String: Any]`, since it'll make stuff easier to compare and use, and introduce more safety

Reply:
i could've sworn i tried and there was a reason we couldn't, but let me try again..